add node shutdown taint

zetaab opened the pull request:
**What this PR does / why we need it**: we need node stopped taint in order to detach volumes immediately without waiting timeout. More info in issue ticket #58635 

**Which issue(s) this PR fixes** 
Fixes #58635

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

zetaab committed:
add node shutdown taint

shutdowned -> stopped

use shutdown everywhere

use patch in taints api call

use notimplemented in clouds use AddOrUpdateTaintOnNode

correct log text

add fake cloud

try to fix bazel

add shutdown tests

add context

zetaab commented:
Questions/comments:

- can we modify Instances interface?
- I need help implementing at least gce/aws part of InstanceShutdownedByProviderID
- the main point why we need this is because of volume detach things. Should that be separate PR or also in this one? After this node label thing is added we should somehow query node taints https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L202 here. If node has shutdown taint, we can skip waiting time. How we should query that? Query node state from apiserver using kubeclient? @jingxu97 
- this needs cloud-controller-manager in use. However, the current situation is that products like kops / openshift are not using cloud-controller-manager. If cloud-controller-manager is not in use, the taints are not added.

yastij commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -239,9 +239,39 @@ func (cnc *CloudNodeController) MonitorNode() {
+					if cloudTaint == nil {
+						taint := v1.Taint{
+							Key:    algorithm.TaintNodeStopped,
+							Value:  "true",
```
Adding the taints means that node is shutdown, no need for a value

yastij commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -239,9 +239,39 @@ func (cnc *CloudNodeController) MonitorNode() {
+						glog.V(4).Infof("Appending stopped taint to node: %s", node.Name)
+						node.Spec.Taints = append(node.Spec.Taints, taint)
+
+						_, err = cnc.kubeClient.CoreV1().Nodes().Update(node)
```
Is this an update or a patch ?

zetaab replied:
patch would be better, but then I must implement more code. For instance https://github.com/zetaab/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L366 is using update as well in this file.

yastij replied:
There's #56350 to fix this, also there was some constrains that made us use update. If you use update you might overwrite Status, which is something we don't want.

yastij commented on pkg/scheduler/algorithm/well_known_labels.go:
```diff
@@ -61,4 +61,9 @@ const (
+	// When node is shutdown in external cloud provider
+	// stopped flag is needed for immediately volume detach
+	// after node comes back, the taint is removed
+	TaintNodeStopped = "node.cloudprovider.kubernetes.io/stopped"
```
We should choose between shutdown or stopped to be consistent

zetaab replied:
maybe shutdown is better word for this

yastij reviewed (commented):
I did a quick pass, I'll do another one tomorrow

zetaab commented:
@yastij check last commit, it is now modified to use patch instead of update.

jhorwit2 commented:
@zetaab Lets please keep the implementation of clouds out of this PR and focus on the API change & node controller changes.

zetaab commented:
@jhorwit2 so you want that I remove that openstack implementation as well? Is there now some code which needs to be removed? Could you comment more specifically. Now this solution is working one in openstack.

jhorwit2 commented:
@zetaab Yeah, for this PR lets focus on the implementation / api changes and save any cloud specific changes for future PR's since not all clouds treat stopped the same (https://github.com/kubernetes/kubernetes/issues/46442). We'll want to track that for each provider and work with the respective cloud providers to update the functionality so it aligns. We should return the [unimplemented error](https://github.com/zetaab/kubernetes/blob/58c41f60560f47cd68a09ebe317d573e0b82dab4/pkg/cloudprovider/cloud.go#L186) and handle it gracefully in the node controller to provide backwards compatibility.

zetaab commented:
@jhorwit2 when you are talking about node controller do you mean this new cloud controller? Because I am not sure should I implement something to old kube-controller? Currently if people do not use cloud controller node monitoring is handled by kube-controller. Should it work there as well or is this cloud controller enough?

yastij commented on pkg/cloudprovider/providers/openstack/openstack_instances.go:
```diff
@@ -129,14 +134,37 @@ func (i *Instances) InstanceExistsByProviderID(providerID string) (bool, error)
 	return true, nil
 }
 
+// InstanceShutdownByProviderID returns true if the instances is in safe state to detach volumes
```
This change shouldn't be included in this PR as @jhorwit2 said

jhorwit2 commented:
> when you are talking about node controller do you mean this new cloud controller?

@zetaab I'm referring to the cloud node controller you made the changes to in this PR already (`pkg/controller/cloud/node_controller.go`)

jhorwit2 commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -239,9 +241,38 @@ func (cnc *CloudNodeController) MonitorNode() {
+
+				if exists {
+					// if node is shutdown add shutdown taint
+					cloudTaint := getTaint(node.Spec.Taints, algorithm.TaintNodeShutdown)
```
You should use [`AddOrUpdateTaintOnNode`](https://github.com/kubernetes/kubernetes/blob/d6325933e1b5ac8a2c0a5a65ab8bd690015f4697/pkg/controller/controller_utils.go#L893) and remove most of this logic including the patch method and undo the changes to `getCloudTaint`.

jsafrane commented:
/ok-to-test

jsafrane commented:
cc: @jingxu97 @gnufied

jdumars commented:
cc: @khenidak

yastij commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -239,9 +246,28 @@ func (cnc *CloudNodeController) MonitorNode() {
 		// from the cloud provider. If node cannot be found in cloudprovider, then delete the node immediately
 		if currentReadyCondition != nil {
 			if currentReadyCondition.Status != v1.ConditionTrue {
+				// we need to check this first to get taint working in similar in all cloudproviders
```
can we have some tests for this ?

zetaab replied:
I can add those when i have time. Hopefully today/tomorrow

zetaab replied:
@yastij tests added now

gmarek commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -271,6 +297,12 @@ func (cnc *CloudNodeController) MonitorNode() {
 
+			} else {
+				// if taint exist remove taint
+				err = controller.RemoveTaintOffNode(cnc.kubeClient, node.Name, node, ShutDownTaint)
```
This means that the semantics or this Taint are somewhat cloudy. 
- It's applied when Node is NotReady and VM is in a given state
- It's removed when Node is Ready

It may be correct, but I guess full negation would be better, i.e.
- It's removed when Node is Ready or VM is not in the given state

zetaab replied:
You are correct that it would be better to update that all the time. However, if we do that we will make request to cloudprovider api every 5 seconds. 99% of these requests are not needed. State is not changed. If we have 100 node clustee it maked 100 api requests every 5 second. Imo too mich because its not really needed

gmarek replied:
Instead of asking about every single VM we can just ask for the list of all VMs and work with that.

jhorwit2 replied:
@zetaab what do you mean we will make it every 5 seconds? Nothing I see Is on such a short interval.

zetaab replied:
@jhorwit2 monitornode function is executed every 5 seconds

jhorwit2 replied:
That seems unnecessarily quick. The KCM marks nodes not ready after 40 seconds by default, so i'm not sure why this would need to be faster than that.

jhorwit2 replied:
@zetaab To your point earlier, this would cause up to 200 api calls every 5 seconds to the cloud provider since you'd not only check for if it's shutdown but also check if it's been deleted for every node that's healthy.

zetaab replied:
I think we could do this whole thing little bit better. If we increase monitornode time, it will mean that removing / adding shutdown taint takes 40seconds+monitornode delay maximum. I mean that if node notready takes 40seconds, it might mean that adding taint takes 40seconds+5seconds maximum currently.

What about if we re-structure this whole feature, if node goes notready state why it cannot push message towards monitornode(using channel maybe)? Then there is (almost) no delay, and we do not need refresh that 24/7 every 5seconds?

zetaab replied:
@jhorwit2 yep 200 maximum per 5seconds with 100 nodes. @gmarek well we could do that. However, I would like to modify then this whole monitornode function. Instead of checking nodes status one by one in cloudprovider, we should make bulk query towards cloudprovider (which input contains all nodes providerids). Then cloudprovider queries providerids statuses from cloudprovider API, and returns three arrays as return to this monitornode function. These arrays are: shutdown, active and removed. Where shutdown means = add taint if taint not exist yet. Active = if taint exist remove taint, removed = delete node

zetaab commented:
note: gce and azure vendor libraries does not support retrieving instance state. AWS+openstack does and I have code for those already. This means that we cannot have gce+azure support for this taint thing before vendor libraries are updated.

zetaab commented:
@jingxu97 @khenidak in issue of this PR you said that taints should be added in node lifecycle controller instead of cloud controller? Do you mean that I should remove taint things from cloud controller and move those to similar place here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L668 (and then implement isShutdown(nodeName) to cloudprovider interface). Or what you think about this PR is it valid and can be approved or need changes?

khenidak commented:
@zetaab 
> ..  in issue of this PR you said that taints should be added in node lifecycle controller instead of cloud controller? Do you mean that I should remove taint things from cloud controller and move those to similar place here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L668 (and then implement isShutdown(nodeName) to cloudprovider interface)

**Yes**

> .. Or what you think about this PR is it valid and can be approved or need changes?

Yes - **Reservation** - not blocking merging this:
5 sec is too quick to check a node complete reboot cycle. Specially if the host is under updates. We are forcing everybody to modify controller manager's `--node-monitor-period` because the default value is too small. 

We have `--TaintNodesByCondition` which logically does the same thing ( i realize this covers only disk/mem/cpu). Maybe it should cover *not running* condition as well. *I will try to find a way to bring this node sig.*

> re Arm/SDK does not have running/shutdown state. 

I think it does instance view have this data:
https://resources.azure.com/subscriptions/{SUB-ID}/resourceGroups/{RG-NAME}/providers/Microsoft.Compute/virtualMachines/{VM-NAME}/instanceView

and the sdk have it [here] (https://github.com/Azure/azure-sdk-for-go/blob/master/arm/compute/virtualmachines.go#L638)

khenidak commented:
Folks - Can we go ahead and merge this to give cloud provider a chance to implement the needed interfaces?

khenidak commented:
@jdumars  for tracking. 1.10 will have new interface for shutdown taints
/sig @kubernetes/sig-azure-misc

zetaab committed:
taint also node controller

fix function

fix gofmt

fix function return value

fix tests

skip notimplemented error

remove factory unused

in openstack we should try to find instanceid from all states instead of ACTIVE, all other cloudproviders do this already

fix tests and lint

fix gofmt

fix nodelifecycletest

fix lint errors

zetaab commented:
Why I modified nodelifecycle controller as well? When we use kubernetes we currently can decide between two node controllers: cloud controller manager or kube controller can handle nodes. The problem in this PR was that modifications was made only to cloud controller manager which is in alpha(?) and its not used yet in many places. However, I decided to add this taint thing to kube controller as well, and it means that we will get taint support faster to volumes. Is this bad thing or what you guys think? Now it is added to all node controllers, it should work in similar way no matter do you use cloud controller manager or kube controller.

zetaab commented:
could you guys review this PR? If this is going to be in 1.10 we must give some time to implement cloudproviders part as well. I can do AWS and openstack but it cannot be done before this is merged.

sjenning commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -240,9 +247,28 @@ func (cnc *CloudNodeController) MonitorNode() {
+				// we need to check this first to get taint working in similar in all cloudproviders
+				// current problem is that shutdown nodes are not working in similar way ie. all cloudproviders
+				// does not delete node from kubernetes cluster when instance it is shutdown see issue #46442
+				exists, err := instances.InstanceShutdownByProviderID(context.TODO(), node.Spec.ProviderID)
```
nit: should `exists` be something like `shutdown`? 

If you do decided to change this, you don't have to change 245/271 below.

jsafrane commented:
/approve 
volume changes

dims commented:
/approve

cloud provider changes

jhorwit2 commented:
/lgtm

jhorwit2 commented:
/assign @deads2k

for controller approval

deads2k commented:
/approve

k82cn commented:
/approve

bsalamat commented:
/approve
for the change in pkg/scheduler

jingxu97 commented:
@zetaab Sorry for sending this late. But I have one concern about it. My understanding is both node_controller in cloudprovider and node_lifecycle_controller checks nodeExistsInCloudProvider and might delete the node. Either controller might delete the node first. For Cinder/Aws, if node shuts down, cloudprovider will return false for nodeExistsInCloudProvider call. So if node_lifecycle_controlle checks first, it will delete the node directly without taint the node. 
Now the problem is that volume manager will depend on the taint to determine whether mounts are gone or not. So if node is deleted without the taint, volume manager will still think the mounts exist.
cc @yujuhong

aleksandra-malinowska commented:
Currently [all Cluster Autoscaling tests](https://k8s-testgrid.appspot.com/sig-autoscaling#gci-gce-autoscaling) are failing due to nodes becoming unready, but never being removed. I've found the following in [kube-controller-manager.log](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling/6055/artifacts/bootstrap-e2e-master/cluster-autoscaler.log) from test run for one such node:

`E0216 01:14:27.084923       1 node_lifecycle_controller.go:686] Error determining if node bootstrap-e2e-minion-group-01b1 shutdown in cloud: failed to get instance ID from cloud provider: instance not found`

This line keeps appearing for the rest of the run (~6h). Looking at the code, it seems the node is stuck in Unready state because of this check: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L684 and the part of the code where it can removed is never reached.

zetaab commented:
@aleksandra-malinowska, right its bug. Here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L685 here we ignore only cloudprovider.NotImplemented error. However, there can be other errors as well which should be ignored https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/util/node/controller_utils.go#L199

zetaab commented:
@jingxu97 I am checking that shutdown thing in node_lifecycle_controller first https://github.com/zetaab/kubernetes/blob/nodetaint/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L682-L714 nodeShutdownInCloudProvider. So it should not be a problem? In both controllers I am doing that. However, this PR is coming back there was one bug that is mentioned above.

justinsb commented:
When we reapply this, let's review it alongside at least some cloudprovider implementation changes:

* I think this interface requires two API calls where previously we needed one, so there may be quota issues.  I'd prefer to see a call that returns another interface that represents the instance state, personally - I think that would help us avoid an explosion of these methods.
* I'd also personally rather see us passing in the api.Node object if we have it, rather than passing in `node.Spec.ProviderID`.  That can help cloudproviders avoid extra API calls - though ProviderID is usually the one we want, we've switched to Node in the past and it's been helpful.
* I'd rather call the method (whether it on a new `NodeState` interface or on the top-level interface) something more explicit, like `ShouldTaintWithShutDownTaint`.
* The really important question... ShutDown or Shutdown :-)

zetaab commented:
@justinsb good points. Actually I was waiting when someone says that we could do this in one interface call instead of two different :) I will post my proposal later today about modifying cloudprovider interface

zetaab commented:
@justinsb @jhorwit2 @jingxu97 what you think about interface like this https://github.com/zetaab/kubernetes/blob/proposal/pkg/cloudprovider/cloud.go#L151-L154 by using that, we could get clearly states in one api query what we need. Exists = no need for action, InstanceShouldRemove = cloud/nodelifecycle controller takes action and will delete node, InstanceShutdown = taint is added. In future we can also add states to this function.

Example what it looks like in openstack https://github.com/zetaab/kubernetes/blob/proposal/pkg/cloudprovider/providers/openstack/openstack_instances.go#L149-L170 

and cloud controller (not maybe perfect yet) https://github.com/zetaab/kubernetes/blob/proposal/pkg/controller/cloud/node_controller.go#L250-L305 

please note that this is work in progress version (I have not even compiled it). Also I will add similar structure to nodelifecycle controller that it works in similar way.

jingxu97 commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -240,9 +247,28 @@ func (cnc *CloudNodeController) MonitorNode() {
+				// current problem is that shutdown nodes are not working in similar way ie. all cloudproviders
+				// does not delete node from kubernetes cluster when instance it is shutdown see issue #46442
+				exists, err := instances.InstanceShutdownByProviderID(context.TODO(), node.Spec.ProviderID)
+				if err != nil && err != cloudprovider.NotImplemented {
```
It looks like we cannot skip here.

jingxu97 commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -240,9 +247,28 @@ func (cnc *CloudNodeController) MonitorNode() {
+					continue
+				}
+
+				if exists {
```
If shutdown state returns without err, taint the node. Otherwise, continue checking

zetaab commented:
@jingxu97  I cannot continue pushing to this PR before it is reverted. However, I did changes like this https://github.com/zetaab/kubernetes/commit/44e63d4e609f274410642934276bf01be7ce7d3d both controllers are now like you mentioned? Then in nodelifecycle controller, I am not calling api anymore to get ProviderID instead I am using node.Spec.ProviderID. This will remove one api call. 

Still we could do this better, but is there sense to do big change? We could combine these two api calls to one https://github.com/zetaab/kubernetes/blob/nodetaint/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L684 and https://github.com/zetaab/kubernetes/blob/nodetaint/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L696 but like I said its big change, and I am not sure do I want to make that. I have already one revert in my CV :)
